### PR TITLE
build: Add recursive board search

### PR DESCRIPTION
**Pull request by jaronkelleher**
Add the ability for boards to be nested into deeper directories.

Signed-off-by: Jaron Kelleher <jkelleher@fb.com>

### CODE REVIEW

File: cmake/app/boilerplate.cmake [hunk 1]
```diff
@@ -341,11 +341,17 @@ foreach(root ${BOARD_ROOT})
...
-    PATHS ${root}/boards/*/*
-    NO_DEFAULT_PATH
-    )
+  # Perform a recursive search underneath the ${root}/boards directory to find a
+  # file named ${BOARD}_defconfig. Once found, set the BOARD_DIR variable to the
+  # parent directory and set the BOARD_DIR_ROOT variable so it can be used later. Once
+  # the directory is found, this search does not need to be performed again.
+  # Note: find_path() does not search recursively, and is not able to perform this task.
+  if(NOT BOARD_DIR)
+    file(GLOB_RECURSE tmp_BOARD_DIR ${root}/boards/*/${BOARD}_defconfig)
```
Comment (galak) [inline cmake/app/boilerplate.cmake:350]: curious if this works ok on windows?

**Comment by galak** [comment]
@carlescufi is there any implication to board name completion support?

**Comment by jaronkelleher** [comment]
> @carlescufi is there any implication to board name completion support?

zephyr is still able to find available boards if an invalid board name is given. Is that what you are referring to?

File: cmake/app/boilerplate.cmake (hunk 1 above)
Comment (jaronkelleher) [reply cmake/app/boilerplate.cmake:350]: It appears to build OK on Windows still. I had someone test this change with a nested board definition on windows and it built successfully.

**Comment by tejlmand** [comment]
> > @carlescufi is there any implication to board name completion support?
> 
> zephyr is still able to find available boards if an invalid board name is given. Is that what you are referring to?

I believe @galak  might be referring to commands, such as `west boards` ?
Will that discover all boards ?
Based on: https://github.com/zephyrproject-rtos/zephyr/blob/8f2d69c4a3e24f4afcef7737cc31d17155a97122/cmake/boards.cmake#L31-L34
I assume it will, but I haven't tested it.

Also, is `cmake -P cmake/boards.cmake` still presenting boards correctly for each arch ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
This is a general comment, not related to the code changes themselves.

Thank you for joining and proposing this enhancement, but this is a bit more than just a build system change, cause it impacts the way boards are organized in Zephyr.

I would therefore like a specific issue that describes the limitaions you see with the current solution, or as minimum, an updated description in this PR of:
- What limitations do you see in current solution
- What are the pros / cons of this extended functionality
- Is this PR intended for Zephyr boards or out of tree boards ?

I have an idea of your intentions, but would like your own description.

Zephyr, and especially number of boards, has grown extensively, so the possibility of better organization of boards could be valuable.
But I would like more people to join such discussion, as well as discussions regarding how to present and describe this new structure.
One benefit of current structure, is that it is very easy to do 
`ls boards/arm/` and see all `arm` based boards.
This will be harder when boards can be further down the folder structure.

The proposed possibility in this PR also requires an update here: https://docs.zephyrproject.org/latest/guides/porting/board_porting.html#create-your-board-directory

**Review by tejlmand — CHANGES_REQUESTED** [review]
some code comments.

**Comment by jaronkelleher** [comment]
> > > @carlescufi is there any implication to board name completion support?
> > 
> > 
> > zephyr is still able to find available boards if an invalid board name is given. Is that what you are referring to?
> 
> I believe @galak might be referring to commands, such as `west boards` ?
> Will that discover all boards ?
> Based on:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/8f2d69c4a3e24f4afcef7737cc31d17155a97122/cmake/boards.cmake#L31-L34
> 
> 
> I assume it will, but I haven't tested it.
> Also, is `cmake -P cmake/boards.cmake` still presenting boards correctly for each arch ?

The boards don't appear in alphabetical order, but they still are all listed in the correct arch section.

```
$ tree boards/x86 -ifF | grep _defconfig
boards/x86/qemu_x86/qemu_x86_64_defconfig
boards/x86/qemu_x86/qemu_x86_coverage_defconfig
boards/x86/qemu_x86/qemu_x86_defconfig
boards/x86/qemu_x86/qemu_x86_nommu_defconfig
boards/x86/qemu_x86/qemu_x86_nopae_defconfig
boards/x86/qemu_x86/qemu_x86_xip_defconfig
# Moved these two boards into nested directories temporarily
boards/x86/tmp1/tmp2/acrn/acrn_defconfig
boards/x86/tmp1/tmp3/minnowboard/minnowboard_defconfig
boards/x86/up_squared/up_squared_32_defconfig
boards/x86/up_squared/up_squared_defconfig
```

This is the output for the x86 section from `cmake -P cmake/boards.cmake`

```
x86:
  qemu_x86_64
  qemu_x86_coverage
  qemu_x86
  qemu_x86_nommu
  qemu_x86_nopae
  qemu_x86_xip
  acrn
  minnowboard
  up_squared_32
  up_squared
```

File: cmake/app/boilerplate.cmake [hunk 2]
```diff
@@ -341,11 +341,17 @@ foreach(root ${BOARD_ROOT})
...
-  find_path(BOARD_DIR
-    NAMES ${BOARD}_defconfig
-    PATHS ${root}/boards/*/*
-    NO_DEFAULT_PATH
-    )
+  # Perform a recursive search underneath the ${root}/boards directory to find a
+  # file named ${BOARD}_defconfig. Once found, set the BOARD_DIR variable to the
+  # parent directory and set the BOARD_DIR_ROOT variable so it can be used later. Once
+  # the directory is found, this search does not need to be performed again.
+  # Note: find_path() does not search recursively, and is not able to perform this task.
```
[the remark this answers is not in the document]
Comment (jaronkelleher) [reply cmake/app/boilerplate.cmake]: Updated the comment with this recommendation

File: cmake/app/boilerplate.cmake [hunk 3]
```diff
@@ -341,11 +341,17 @@ foreach(root ${BOARD_ROOT})
...
+  if(NOT BOARD_DIR)
+    file(GLOB_RECURSE tmp_BOARD_DIR ${root}/boards/*/${BOARD}_defconfig)
+    get_filename_component(BOARD_DIR ${tmp_BOARD_DIR} DIRECTORY)
```
[the remark this answers is not in the document]
Comment (jaronkelleher) [reply cmake/app/boilerplate.cmake]: Updated to recommended change

File: cmake/app/boilerplate.cmake [hunk 4]
```diff
@@ -341,11 +341,17 @@ foreach(root ${BOARD_ROOT})
       message("Board alias ${BOARD_ALIAS} is hiding the real board of same name")
     endif()
   endif()
-  find_path(BOARD_DIR
-    NAMES ${BOARD}_defconfig
-    PATHS ${root}/boards/*/*
-    NO_DEFAULT_PATH
-    )
+  # Perform a recursive search underneath the ${root}/boards directory to find a
```
[the remark this answers is not in the document]
Comment (jaronkelleher) [reply cmake/app/boilerplate.cmake:345]: Updated the BOARD_ALIAS logic to function in the same way

**Comment by carlescufi** [comment]
> > > @carlescufi is there any implication to board name completion support?
> > 
> > 
> > zephyr is still able to find available boards if an invalid board name is given. Is that what you are referring to?
> 
> I believe @galak might be referring to commands, such as `west boards` ?
> Will that discover all boards ?
> Based on:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/8f2d69c4a3e24f4afcef7737cc31d17155a97122/cmake/boards.cmake#L31-L34
> 
> 
> I assume it will, but I haven't tested it.
> Also, is `cmake -P cmake/boards.cmake` still presenting boards correctly for each arch ?

As far as the in-tree boards is concerned, this patch has no impact on the functionality, both when invoked via `cmake -P cmake/boards.cmake` or `west build -b <TAB>`.  So no issues there.

**Comment by carlescufi** [comment]
@jaronkelleher would you mind responding to @tejlmand's comments [here](https://github.com/zephyrproject-rtos/zephyr/pull/29297#pullrequestreview-512505519)?

**Comment by mbolivar-nordic** [comment]
> @jaronkelleher would you mind responding to @tejlmand's comments [here](https://github.com/zephyrproject-rtos/zephyr/pull/29297#pullrequestreview-512505519)?

I agree an issue where we can discuss would be good. This is one of those changes that feels small but probably isn't.

I also wonder how it interacts with https://github.com/zephyrproject-rtos/zephyr/issues/26522.

**Comment by jaronkelleher** [comment]
> > @jaronkelleher would you mind responding to @tejlmand's comments [here](https://github.com/zephyrproject-rtos/zephyr/pull/29297#pullrequestreview-512505519)?
> 
> I agree an issue where we can discuss would be good. This is one of those changes that feels small but probably isn't.
> 
> I also wonder how it interacts with #26522.

I'll have an issue posted by EOW to continue this discussion

**Comment by jaronkelleher** [comment]
> > > @jaronkelleher would you mind responding to @tejlmand's comments [here](https://github.com/zephyrproject-rtos/zephyr/pull/29297#pullrequestreview-512505519)?
> > 
> > 
> > I agree an issue where we can discuss would be good. This is one of those changes that feels small but probably isn't.
> > I also wonder how it interacts with #26522.
> 
> I'll have an issue posted by EOW to continue this discussion

Time really got away from me here, sincerest apologies to everyone. I've posted an issue for further discussion: #30612

**Pull request closed without merging**